### ISSUE TITLE
A new term in the likelihood for the template analysis

### ISSUE THREAD

**Issue by micheledoro** [gammapy/gammapy#4310]
For the dSph KSP we would like to test a feature and we are not sure if it's possible to do it now or it is a feature that can be implemented.

We are using a template analysis. Our sky model, besides the poissonian noise, has a fluctuation term related on the uncertainty of the morphology of the signal. This morphology is peaked at the center and steeply falling (a cusp), _but with a different uncertainty for each radial bin_ We were curious whether or how we can input such an uncertainty in the spatial model.

Furthermore, the PDF of this uncertainty, bin by bin, is not a symmetric gaussian but can be better reported as a  triangular PDF (although other PDF can be investigated)

Can anyone enlighten us whether using bin by bin radial PDF is the spatial model is possible?

**Comment by adonath** [comment]
Thanks @micheledoro! We are working on a prior mechanism for parameters and models in Gammapy. Something like you described should become possible by defining a custom prior on the level of the `TemplateSpatialModel`.

**Comment by micheledoro** [comment]
Hi @adonath, thanks! So I guess it's not available yet as a method right? In this case we will be happy to test it when ready!

**Comment by adonath** [comment]
> Hi @adonath, thanks! So I guess it's not available yet as a method right? In this case we will be happy to test it when ready!

No, the feature will likely come in v1.1 (spring) or v1.2 (autumn)...I'll report on the progress in the issue, so you get notified.

**Comment by registerrier** [comment]
Hi @micheledoro . Is this still something you want to achieve? The `Prior` and `Sampler` should help. Are there other features missing?

**Comment by micheledoro** [comment]
Hi @registerrier. This was a long time ago. I had to re-read the discussion. I did not get from the thread if you implemented this or not. This = uncertainties on the spatial term of the signal model that change with a user-defined formula bin-by-bin.

Concrete example. A DM profile decrease with say r^-2, but the uncertainty on the value increase with r^1.1. (an example)

**Comment by registerrier** [comment]
OK. 

Currently you can add a `Prior` on any parameter. This means that you can represent the uncertainty if there is an explicit nuisance parameter it relates to.

If the profile is defined as a custom model with `n` parameters giving the flux `F_i` at radius `r_i` you can add a prior on the value of `F_i` representing its uncertainty. This means that you have to explicitly introduce and fit or marginalize over the nuisance parameters representing the actual profile.  

Conversely, you could also have a spatial model with an explicit parameter `alpha` such that `F(r) \propto r^-alpha` and introduce a prior on the value of `alpha` to marginalize over its value. 

What is not clear to me is how to properly introduce a position dependent penalty on the likelihood independently of a (set of) parameter(s).

**Comment by micheledoro** [comment]
Thanks. Very interesting. The project in which I had planned to use this, is over now, but I will certainly keep in mind to have a look at this feature (for dark matter searches). @giacomodamico24 remind me to discuss this with you next time we meet.

**Comment by QRemy** [comment]
Closing this for now, feel free to re-open if you find an use case that cannot be addressed with the `Prior` approach described in previous comments.
